Content.Server/Fluids/EntitySystems/AbsorbentSystem.cs

```diff
@@ -117,30 +117,43 @@ private bool TryTransferAbsorber(EntityUid user, EntityUid used, EntityUid targe
         if (!_solutionSystem.TryGetRefillableSolution(target, out var refillableSolution, refillable: refillable))
             return false;
 
-        if (refillableSolution.Volume <= 0)
+        if (refillableSolution.Volume < 0)
         {
             var msg = Loc.GetString("mopping-system-target-container-empty", ("target", target));
             _popups.PopupEntity(msg, user, user);
             return false;
         }
 
+        var transferAmount = component.PickupAmount / PuddleSystem.EvaporationReagentRatio -
+            absorberSoln.GetReagentQuantity(PuddleSystem.EvaporationReagent);
+
+        transferAmount = transferAmount > 0 ? transferAmount : 0;
+
         // Remove the non-water reagents.
         // Remove water on target
         // Then do the transfer.
-        var nonWater = absorberSoln.SplitSolutionWithout(component.PickupAmount, PuddleSystem.EvaporationReagent);
+        var water = refillableSolution.RemoveReagent(PuddleSystem.EvaporationReagent, transferAmount);
+
+        // Reads from refillableSolution.AvailableVolume have to
+        // be done after possibly removing some water
+        if (refillableSolution.AvailableVolume == FixedPoint2.Zero){
+            _popups.PopupEntity(Loc.GetString("mopping-system-full", ("used", target)), user, user);
```

Comment:
Why is the reagent being removed before checking for the available volume? The water also isn't transferred it just gets deleted.

Reply:
There's no reagents deletion, at least not in the version I tested it.
The case where it seems it might be losing some reagents, it's actually inside a conditional which means "is the mop bucket full?" so if it is full, it couldn't possibly have had some water just removed from it, hence the assert instead of doing something with the "water".
There's an Assert to prevent people evolving the code from breaking that "contract", but if you want I can make it into a more explicit error. It's just not going to be reached by this version.

Reply:
In other words, it will only state that a mop bucket is full after it has no water. If it didn't have any AvailableVolume but still had some water it would transfer that and probably fill the mop bucket with puddle volume from the mop.